packages/zk-certificates/lib/keyManagement.ts

```diff
@@ -2,22 +2,51 @@
 import {
   eddsaKeyGenerationMessage,
   eddsaPrimeFieldMod,
+  EddsaPrivateKey,
 } from '@galactica-net/galactica-types';
 import createBlakeHash from 'blake-hash';
 import { Buffer } from 'buffer';
+import { Eddsa } from 'circomlibjs';
 import { Signer } from 'ethers';
 import { Scalar, utils } from 'ffjavascript';
 
 /**
  * Generates the eddsa private key from the ethereum private key signing a fixed message.
  *
  * @param signer - Ethers signer.
- * @returns The eddsa private key.
+ * @returns The eddsa private key as buffer.
  */
 export async function getEddsaKeyFromEthSigner(
   signer: Signer,
-): Promise<string> {
-  return signer.signMessage(eddsaKeyGenerationMessage);
+): Promise<EddsaPrivateKey> {
+  // use signature as entropy input so that the EdDSA key can be derived from the Ethereum private key
+  const signature = await signer.signMessage(eddsaKeyGenerationMessage);
+  return getEddsaKeyFromEntropy(signature.slice(2));
+}
+
+/**
+ * Generates the EdDSA private key, following https://www.rfc-editor.org/rfc/rfc8032#section-5.1.5 .
+ *
+ * @param entropy - Random entropy to generate key from as hex string.
+ * @returns The eddsa private key as buffer.
+ */
+export function getEddsaKeyFromEntropy(entropy: string): EddsaPrivateKey {
+  let source = entropy;
+  if (entropy.startsWith('0x')) {
+    source = entropy.slice(2);
+  }
+  if (source.length < 64) {
```

Comment:
The check if it is a hex string is directly after. The order should not matter, but I'll swap it to be simpler to read.

Reply:
The order does matter, because [`string.length`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/length) returns not bytes. Code units are not the same as bytes. Check [§3.9 "Unicode Encoding Forms", D77](https://www.unicode.org/versions/Unicode15.0.0/UnicodeStandard-15.0.pdf) for reference.